### PR TITLE
no-std for ff-fft and groth16

### PR DESCRIPTION
**Pull request by debris**
built on top of pr #76

### CODE REVIEW

File: ff-fft/src/domain_seq.rs [hunk 1]
```diff
@@ -0,0 +1,422 @@
+//! This module contains an `EvaluationDomain` abstraction for
```
Comment (Pratyush) [inline ff-fft/src/domain_seq.rs]: Do we need this file? I think we can just `#cfg[feature = "std"]` out anything that depends on `rayon`.

File: ff-fft/src/polynomial/dense.rs [hunk 2]
```diff
@@ -104,11 +106,24 @@ impl<F: Field> DensePolynomial<F> {
         }
         assert_eq!(powers_of_point.len(), self.coeffs.len());
         let zero = F::zero();
-        powers_of_point
```
Comment (Pratyush) [inline ff-fft/src/polynomial/dense.rs:107]: Let's do something similar to here: https://github.com/scipr-lab/zexe/blob/8a9343ad03e24cf278598e19c9a7b928b3ab729e/algebra/src/msm/fixed_base.rs#L80

This avoid duplicating the rest of the iterator chain.

File: ff-fft/src/polynomial/dense.rs [hunk 3]
```diff
@@ -143,10 +158,22 @@ impl<F: PrimeField> DensePolynomial<F> {
     pub fn mul_by_vanishing_poly(&self, domain: EvaluationDomain<F>) -> DensePolynomial<F> {
         let mut shifted = vec![F::zero(); domain.size()];
         shifted.extend_from_slice(&self.coeffs);
-        shifted
```
Comment (Pratyush) [inline ff-fft/src/polynomial/dense.rs:146]: Let's do something similar to here: https://github.com/scipr-lab/zexe/blob/8a9343ad03e24cf278598e19c9a7b928b3ab729e/algebra/src/msm/fixed_base.rs#L80

This avoid duplicating the rest of the iterator chain.

**Review by Pratyush — CHANGES_REQUESTED** [review]
Cool thanks! I merged #76 and also added a new PR (#86) that simplifies the code in `domain.rs`, so if you could rebase that'd be great.

File: ff-fft/src/polynomial/dense.rs (hunk 3 above)
Comment (debris) [reply ff-fft/src/polynomial/dense.rs:146]: fixed

File: ff-fft/src/polynomial/dense.rs (hunk 2 above)
Comment (debris) [reply ff-fft/src/polynomial/dense.rs:107]: in this case the iterator chain is not duplicated, cause `fold/reduce` function from `rayon` has a different `signature` than `fold` from `std` and it requires us to pass `Fn` as accumulator

[rayon-docs#method.fold](https://docs.rs/rayon/1.3.0/rayon/iter/trait.ParallelIterator.html#method.fold)

[std-docs#method.fold](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold)

File: ff-fft/src/domain_seq.rs (hunk 1 above)
Comment (debris) [reply ff-fft/src/domain_seq.rs]: you are right, after simplifying the code in #86 it is much easier to just add `#cfg[feature = "std"]` everywhere it is needed

Comment (Pratyush) [reply ff-fft/src/domain_seq.rs]: To clarify, I meant that we can use `#cfg[feature = "std"]` on the original `domain.rs`.

**Comment by debris** [comment]
I think it's ready for another review :)

**Comment by Pratyush** [comment]
Hmm it looks generally good. I noticed that there's a lot of code duplication due to the `cfg`s. Maybe we can define a macro like `cfg_iter` that does something like:
```rust
macro_rules! cfg_iter {
    ($e:expr) => {
        #[cfg(feature = "parallel")]
		let result = $e.par_iter()
		#[cfg(not(feature = "parallel"))]
		let result = $e.iter();
		result
    }
}
```
This can then be used something like ``` let result = cfg_iter!(array_to_be_iterated_over).zip(other_array)...```.

**Comment by debris** [comment]
Added 3 new macros `cfg_iter`, `cfg_iter_mut` and `cfg_into_iter` :)

**Comment by Pratyush** [comment]
Awesome! This is great! I’ll merge once CI passes :)

**Pull request merged**